fuse_models/src/odometry_2d_publisher.cpp

```diff
@@ -75,7 +75,8 @@ void Odometry2DPublisher::onInit()
 
   params_.loadFromROS(private_node_handle_);
 
-  if (params_.world_frame_id == params_.map_frame_id)
+  const auto frame_id = params_.invert_tf ? params_.base_link_output_frame_id : params_.world_frame_id;
+  if (frame_id == params_.map_frame_id)
```

Comment:
Regardless of the state of the `invert_tf` flag, the odom message will use the world_frame_id as the parent frame and base_link_output_frame_id as the child frame. That makes me think this check should continue to be:
```
if (params_.world_frame_id == params_.map_frame_id)
``` 
regardless of the `invert_tf` flag.

What I'm having trouble working out is what should this do?
```
world_frame_id: map
base_link_output_frame_id: base_link
invert_tf: true
```
The odometry message will use (parent frame "map", child frame "base_link").
Without the `invert_tf` flag, tf would publish (parent frame "map", child frame "odom").
Does that mean with the `invert_tf`, tf should publish (parent frame "odom", child frame "map")? But I have no idea why anyone would ever want that. Do you have a use-case in mind?

But with the code as-is, I do not think it is computing a valid transform. The base_link->odom lookup is not modified by the `invert_tf` flag, so it is always base_link->odom. The transform `trans` is modified by the `invert_tf` flag, so `trans` is really base_link->map. And the two get multiplied together here (https://github.com/locusrobotics/fuse/blob/devel/fuse_models/src/odometry_2d_publisher.cpp#L468):
```
base_link->map * base_link->odom
```
which I do not believe yields anything meaningful.

If publishing odom->map is the expected output, I think it would be easier to handle the invert logic just before the broadcast. So all of the code would remain identical, and right before this (https://github.com/locusrobotics/fuse/blob/devel/fuse_models/src/odometry_2d_publisher.cpp#L481), you add an `if(params_.invert_tf)` block that computes the inverse transform. It's a little annoying because you have to bounce back and forth between tf and geometry_msgs. There's probably a more clever way to do this with fewer translations. Converting the tf lookup results to a tf Transform maybe? And delaying the conversion from tf Transform to geometry_msgs until the very end?

Reply:
With `invert_tf` enabled I do `std::swap(frame_id, child_frame_id);`, so the block `if (frame_id == params_.map_frame_id)` compares `base_link` with `map`, so it doesn't compute `base_link->map * base_link->odom`.

The output in that case is simply `base_link->map`.

The whole purpose here is to avoid calling `lookupTransform()`, i.e. skip the need to broadcast wrt `odom` due to TF tree limitations.

Reply:
I feel like what you are saying is, if the "invert_tf" flag it true, we should not create the tf listener or do the tf lookup. Can we just say that in the code, instead of swapping frame names around?

```
if (!params_.invert_tf && params_.world_frame_id == params_.map_frame_id)
```

Reply:
Exactly, that's the intend. Changed, so it's easier to read it in the code. :+1: